platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/crypto/EnhancedKeyStoreLoader.java

```diff
@@ -1165,4 +1175,340 @@ private interface AddressBookCallback {
+
+        // create PEM files for signing keys and certs.
+        long errorCount = extractPrivateKeysAndCertsFromPfxFiles(pfxPrivateKeys, pfxCertificates);
+
+        if (errorCount == 0) {
+            // validate only when there are no errors extracting pem files.
+            errorCount = validateKeysAndCertsAreLoadableFromPemFiles(pfxPrivateKeys, pfxCertificates);
+        }
+
+        if (errorCount > 0) {
+            // roll back due to errors.
+            // this deletes any pem files created, but leaves the agreement keys deleted.
+            logger.error(STARTUP.getMarker(), "Due to {} errors, reverting pem file creation.", errorCount);
+            rollBackSigningKeysAndCertsChanges(pfxPrivateKeys, pfxCertificates);
+        } else {
+            // cleanup pfx files by moving them to sub-directory
+            cleanupMovePfxFilesToSubDirectory();
+        }
+
+        logger.info(STARTUP.getMarker(), "Completed migration of key store.");
+        return this;
+    }
+
+    /**
+     * Delete any agreement keys from the key store directory.
+     */
+    private void deleteAgreementKeys() {
+        // delete any agreement keys of the form a-*
+        final File[] agreementKeyFiles = keyStoreDirectory.toFile().listFiles((dir, name) -> name.startsWith("a-"));
+        if (agreementKeyFiles != null) {
+            for (File agreementKeyFile : agreementKeyFiles) {
+                if (agreementKeyFile.isFile()) {
+                    try {
+                        Files.delete(agreementKeyFile.toPath());
+                        logger.debug(STARTUP.getMarker(), "Deleted agreement key file {}", agreementKeyFile.getName());
+                    } catch (IOException e) {
+                        logger.error(
+                                ERROR.getMarker(),
+                                "Failed to delete agreement key file {}",
+                                agreementKeyFile.getName());
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Extracts the private keys and certificates from the PFX files and writes them to PEM files.
+     *
+     * @param pfxPrivateKeys  the map of extracted private keys.
+     * @param pfxCertificates the map of extracted certificates.
+     * @return the number of errors encountered during the extraction process.
+     * @throws KeyStoreException   if the underlying method calls throw this exception.
+     * @throws KeyLoadingException if the underlying method calls throw this exception.
+     */
+    private long extractPrivateKeysAndCertsFromPfxFiles(
+            final Map<NodeId, PrivateKey> pfxPrivateKeys, final Map<NodeId, Certificate> pfxCertificates)
+            throws KeyStoreException, KeyLoadingException {
+        final KeyStore legacyPublicStore = resolveLegacyPublicStore();
+        AtomicLong errorCount = new AtomicLong(0);
+
+        iterateAddressBook(addressBook, (i, nodeId, address, nodeAlias) -> {
+            if (isLocal(address)) {
```

Comment:
This lambda is very long, in an already long method. Consider extracting into a static method

Reply:
the lambda is the only logic in the method containing it.  There would not be any code improvement from doing that.  Doing the embedded lambda is necessary to get the error count given the API for calling the lambda.